pkg/cgroups/cgroupblkio.go

```diff
@@ -0,0 +1,168 @@
+	}
+}
+
+// NewOciRateDeviceParameters creates new OciRateDeviceParameters instance.
+func NewOciRateDeviceParameters() OciRateDeviceParameters {
+	return OciRateDeviceParameters{
+		Major: -1,
+		Minor: -1,
+		Rate:  -1,
+	}
+}
+
+// GetBlkioDir returns the cgroups blkio controller directory.
+func GetBlkioDir() string {
+	return blkioCgroupDir
+}
+
+// SetBlkioParameters writes OCI BlockIO parameters to files in cgroups blkio contoller directory.
+func SetBlkioParameters(cgroupsDir string, blockIO OciBlockIOParameters) error {
+	log.Debug("configuring cgroups blkio controller in directory %#v with parameters %+v", cgroupsDir, blockIO)
+	var errors *multierror.Error
+	if blockIO.Weight >= 0 {
+		errors = multierror.Append(errors, writeToFileInDir(cgroupsDir, blkioWeightFiles, strconv.FormatInt(blockIO.Weight, 10)))
+	}
+	for _, weightDevice := range blockIO.WeightDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioWeightDeviceFiles, weightDevice.Major, weightDevice.Minor, weightDevice.Weight))
+	}
+	for _, rateDevice := range blockIO.ThrottleReadBpsDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleReadBpsFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	for _, rateDevice := range blockIO.ThrottleWriteBpsDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleWriteBpsFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	for _, rateDevice := range blockIO.ThrottleReadIOPSDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleReadIOPSFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	for _, rateDevice := range blockIO.ThrottleWriteIOPSDevice {
+		errors = multierror.Append(errors, writeDevValueToFileInDir(cgroupsDir, blkioThrottleWriteIOPSFiles, rateDevice.Major, rateDevice.Minor, rateDevice.Rate))
+	}
+	return errors.ErrorOrNil()
+}
+
+// writeDevValueToFileInDir writes MAJOR:MINOR VALUE to the first existing file under baseDir
+func writeDevValueToFileInDir(baseDir string, filenames []string, major, minor, value int64) error {
+	content := fmt.Sprintf("%d:%d %d", major, minor, value)
+	return writeToFileInDir(baseDir, filenames, content)
+}
+
+// writeToFileInDir writes content to the first existing file in the list under baseDir.
+func writeToFileInDir(baseDir string, filenames []string, content string) error {
+	var errors *multierror.Error
+	// Returns list of errors from writes, list of single error due to all filenames missing or nil on success.
+	for _, filename := range filenames {
+		filepath := baseDir + "/" + filename
+		err := writeToFile(filepath, content)
+		if err == nil {
+			return nil
+		}
+		errors = multierror.Append(errors, err)
+	}
+	err := errors.ErrorOrNil()
+	if err != nil {
+		return fmt.Errorf("could not write content %#v to any of files \"%s\": %w", content, strings.Join(filenames, "\", \""), err)
```

Comment:
instead of `\"%s\"` you can use %q. Also, instead of `strings.Join(filenames)` you can just print `%v` of filenames.

Reply:
For readability of CRI-RM output, I'd prefer current error message formatting:
```
could not write content "1:0 100" to any of files "blkio.bfq.weight_device", "blkio.weight_device": ...
```

over `%q` + `filenames`:
```
could not write content "1:0 100" to any of files ["blkio.bfq.weight_device" "blkio.weight_device"]: ...
```

or `%v` + `filenames`:
```
could not write content "1:0 100" to any of files [blkio.bfq.weight_device blkio.weight_device]: ...
```

Please let me know if you'd still like to see either of the latter messages rather than the first, and I'll simplify the code.

Reply:
doing manual quoting `\"%s\"` instead of `%q` from system library doesn't make any sense.
Having brackets around list of files it might be a bit less readable for end-user, but a lot more readable for developer of the code, with less chances to miss escaping of quote characters.

`
return fmt.Errorf("could not write content %#v to any of files %q: %w", content, filenames, err)
`
is more readable than
`
return fmt.Errorf("could not write content %#v to any of files \"%s\": %w", content, strings.Join(filenames, "\", \""), err)
`

Reply:
Ok, I switched to `%q` and removed `Join()`. PR updated.